### ISSUE TITLE
Dead for Parallel Training Using Horovod for Acceleration!

### ISSUE THREAD

**Issue by MagicFrogSJTU** [ultralytics/yolov5#177]
## 🚀 Feature
Training is slow in multi-gpu mode with centerized parameter server, as currently implemented in the form of torch.nn.parallel.DistributedDataParallel.
Discenterized mode like horovod should be better. Plus, I found that most of the time the gpus are waiting for data feeding. Data parallel will be also boost using horovod.
We need horovod!

**Comment by glenn-jocher** [comment]
@MagicFrogSJTU feel free to submit a PR with the changes you had in mind.

**Comment by MagicFrogSJTU** [comment]
I will submit a PR once it's done! But now I am being stuck at reproducing.
It seems like pycoco issues. I will further check on that. 

---------------------------------
I ran the provided docker image locally, and did inference using test.py and provided pretrained model checkpoints.
``bash
python test.py --weights yolov5s.pt --img 736 --conf 0.001 --data data/coco.yaml
``
The problem is, I got a lower performance, approximately -1 mAP.
I got mAP@.5/mAP@.5:.95=55.0/34.9. 
But the perf claimed in README.md is 55.8/36.6.

**Comment by MagicFrogSJTU** [comment]
> @MagicFrogSJTU feel free to submit a PR with the changes you had in mind.

The DDP implementation seems already in the repos, which I neglected before.
But the speed was unexpectedly slow. I am currently working on that

**Comment by MagicFrogSJTU** [comment]
I have probably found the reason of this issue.
#116
The original implementation
`` python
        dist.init_process_group(backend='nccl',  # distributed backend
                                init_method='tcp://127.0.0.1:9999',  # init method
                                world_size=1,  # number of nodes
                                rank=0)  # node ran
``
actually doesn't run DDP mode directly. It is equal to DP.
It is because the word_size is set to 1, which means there is only one process triggered. In this case, given multiple gpus, the process just utilizes them in a data-parallel way.
Therefore, the rank 0 would consume more gpu memories as in normal DP mode.

**Comment by MagicFrogSJTU** [comment]
> I have probably found the reason of this issue.
> #116
> The original implementation
> `python dist.init_process_group(backend='nccl', # distributed backend init_method='tcp://127.0.0.1:9999', # init method world_size=1, # number of nodes rank=0) # node ran`
> actually doesn't run DDP mode directly. It is equal to DP.
> It is because the word_size is set to 1, which means there is only one process triggered. In this case, given multiple gpus, the process just utilizes them in a data-parallel way.
> Therefore, the rank 0 would consume more gpu memories as in normal DP mode.

@glenn-jocher 
The original DDP implementation should be corrected. It is now acting as DP.

**Comment by glenn-jocher** [comment]
@MagicFrogSJTU ok, sounds good! Can you submit a PR please with your proposed changes?

To reproduce a final mAP you should run test.py directly, as pycocotools always returns slightly higher mAP than our internal repo calculations, i.e.:

`python test.py --data coco.yaml --weights yolov5s.pt`

Any other sources of differences may be due to single-GPU (readme results) vs multi-gpu, which may need syncbatchnorm layers for best results.

**Comment by MagicFrogSJTU** [comment]
> @MagicFrogSJTU ok, sounds good! Can you submit a PR please with your proposed changes?
> 
> To reproduce a final mAP you should run test.py directly, as pycocotools always returns slightly higher mAP than our internal repo calculations, i.e.:
> 
> `python test.py --data coco.yaml --weights yolov5s.pt`
> 
> Any other sources of differences may be due to single-GPU (readme results) vs multi-gpu, which may need syncbatchnorm layers for best results.

I will submit a PR as soon as possible!
Before that, I have to finish a complete training procedure to see if it works as expected!

**Comment by glenn-jocher** [comment]
@MagicFrogSJTU great, thank you! Also before submitting the PR please make sure single-GPU operates without errors also in your updated code (i.e. just train a few batches to check).

**Comment by glenn-jocher** [comment]
@MagicFrogSJTU by the way, you can specify devices for training as in these examples:

```
python train.py --device cpu
python train.py --device 0
python train.py --device 4,5,6,7
```

**Comment by MagicFrogSJTU** [comment]
> @MagicFrogSJTU by the way, you can specify devices for training as in these examples:
> 
> ```
> python train.py --device cpu
> python train.py --device 0
> python train.py --device 4,5,6,7
> ```

Of course.
are there any early official performances (for example, 10 epochs ) to align?

**Comment by glenn-jocher** [comment]
@MagicFrogSJTU yes, there are unit tests we run typically before pushing changes. You can run them with single-gpu training and then double-gpu training for example. The tests pass if all runs are executed without error.

```bash
# Unit tests
%%shell
rm -rf yolov5 && git clone https://github.com/ultralytics/yolov5 && cd yolov5
export PYTHONPATH="$PWD" # to run *.py. files in subdirectories
pip install -r requirements.txt onnx
python3 -c "from utils.google_utils import *; gdrive_download('1n_oKgR81BJtqk75b00eAjdv03qVCQn2f', 'coco128.zip')" && mv ./coco128 ../
for x in yolov5s yolov5m yolov5l yolov5x # models
do
  python train.py --weights $x.pt --cfg $x.yaml --epochs 4 --img 320 --device 0  # train
  for di in 0 cpu # inference devices
  do
    python detect.py --weights $x.pt --device $di  # detect official
    python detect.py --weights weights/last.pt --device $di  # detect custom
    python test.py --weights $x.pt --device $di # test official
    python test.py --weights weights/last.pt --device $di # test custom
  done
  python models/yolo.py --cfg $x.yaml # inspect
  python models/export.py --weights $x.pt --img 640 --batch 1 # export
done
```

**Comment by MagicFrogSJTU** [comment]
> @MagicFrogSJTU yes, there are unit tests we run typically before pushing changes. You can run them with single-gpu training and then double-gpu training for example. The tests pass if all runs are executed without error.
> 
> ```shell
> # Unit tests
> %%shell
> rm -rf yolov5 && git clone https://github.com/ultralytics/yolov5 && cd yolov5
> export PYTHONPATH="$PWD" # to run *.py. files in subdirectories
> pip install -r requirements.txt onnx
> python3 -c "from utils.google_utils import *; gdrive_download('1n_oKgR81BJtqk75b00eAjdv03qVCQn2f', 'coco128.zip')" && mv ./coco128 ../
> for x in yolov5s yolov5m yolov5l yolov5x # models
> do
>   python train.py --weights $x.pt --cfg $x.yaml --epochs 4 --img 320 --device 0  # train
>   for di in 0 cpu # inference devices
>   do
>     python detect.py --weights $x.pt --device $di  # detect official
>     python detect.py --weights weights/last.pt --device $di  # detect custom
>     python test.py --weights $x.pt --device $di # test official
>     python test.py --weights weights/last.pt --device $di # test custom
>   done
>   python models/yolo.py --cfg $x.yaml # inspect
>   python models/export.py --weights $x.pt --img 640 --batch 1 # export
> done
> ```

Nice!
Anyway, I would make sure that the first 10 epoch would generate similar system performance for both single-gpu and DDP.

**Comment by MagicFrogSJTU** [comment]
> > @MagicFrogSJTU yes, there are unit tests we run typically before pushing changes. You can run them with single-gpu training and then double-gpu training for example. The tests pass if all runs are executed without error.
> > ```shell
> > # Unit tests
> > %%shell
> > rm -rf yolov5 && git clone https://github.com/ultralytics/yolov5 && cd yolov5
> > export PYTHONPATH="$PWD" # to run *.py. files in subdirectories
> > pip install -r requirements.txt onnx
> > python3 -c "from utils.google_utils import *; gdrive_download('1n_oKgR81BJtqk75b00eAjdv03qVCQn2f', 'coco128.zip')" && mv ./coco128 ../
> > for x in yolov5s yolov5m yolov5l yolov5x # models
> > do
> >   python train.py --weights $x.pt --cfg $x.yaml --epochs 4 --img 320 --device 0  # train
> >   for di in 0 cpu # inference devices
> >   do
> >     python detect.py --weights $x.pt --device $di  # detect official
> >     python detect.py --weights weights/last.pt --device $di  # detect custom
> >     python test.py --weights $x.pt --device $di # test official
> >     python test.py --weights weights/last.pt --device $di # test custom
> >   done
> >   python models/yolo.py --cfg $x.yaml # inspect
> >   python models/export.py --weights $x.pt --img 640 --batch 1 # export
> > done
> > ```
> 
> Nice!

**Comment by MagicFrogSJTU** [comment]
That was a click accident, my bad.

**Comment by MagicFrogSJTU** [comment]
@glenn-jocher 
What's the purpose of using ModelEMA.update_attr() as in 290line in train.py?
To be clear, what attrs are needed? I recommend that update specific attr rather than all of them. Because when you use DDP, or something else, to wrap model, lots os attrs will be introduced, some of which may be not pickable.
It is making things complicated in DDP model when trying to save the model.

**Comment by glenn-jocher** [comment]
@MagicFrogSJTU model attributes need to be saved along with the model, for example model.names is a list of class names used during inference and testing for displaying results and annotating images with the results.

**Comment by MagicFrogSJTU** [comment]
@glenn-jocher 
I have implemented the DDP, and run a entire 4-GPU training procedure for v5s model.
However, the performance is lower.

| model       | mAP@0.5:0.95 | mAP@0.5 |
| -- | -- | -- |
| should be |  36.6 |  55.8 |
| my DDP |   34.440 | 53.513  |

The reason of performance deterioration is still unknown.
I found that the loss is summed over a batch. Given that the ring-all-redue in DDP does mean reduction,  I multiplied the loss of each process with world_size(). This could make perf higher, but still lower than expected.
As the implementation of original compute_loss function is complecated, I wonder, if you know, whether there is something special in the compute_loss() that needs special treatment?

**Comment by MagicFrogSJTU** [comment]
@glenn-jocher

**Comment by MagicFrogSJTU** [comment]
I have checked the compute_loss() thoroughly and didn't found anything surprised.
Meanwhile, I fixed another dataset inconsistency problem. The training is now runing. Let's see if it works this time..

After tons of time spent on checking the code, I am quite sure that now the only last possible problem here is the BN problem.
SyncBN is probably needed.

**Comment by glenn-jocher** [comment]
@MagicFrogSJTU for best multi-gpu performance we'd need to replace BatchNorm with SyncBatchNorm. Is your multi-gpu implementation faster than the default?

**Comment by glenn-jocher** [comment]
@MagicFrogSJTU also one thing to keep in mind is that if you change batch-size you will get different training results. In general the results should improve up to batch-size 64, past that I'm not sure. I trained yolov5s up to batch-size 80 with similar results to batch-size 64, but for example if you have 4 GPUs and you use batch-size 256 you will likely see some changes in mAP, since the training is doing 4X fewer updates.

This means that the largest models will probably benefit the most from multi-gpu as they can grow their batch size towards 64, while the smallest model is already capable of batch size 64 with single-gpu.

**Comment by glenn-jocher** [comment]
For syncbn see https://pytorch.org/docs/master/generated/torch.nn.SyncBatchNorm.html

**Comment by MagicFrogSJTU** [comment]
> @MagicFrogSJTU also one thing to keep in mind is that if you change batch-size you will get different training results. In general the results should improve up to batch-size 64, past that I'm not sure. I trained yolov5s up to batch-size 80 with similar results to batch-size 64, but for example if you have 4 GPUs and you use batch-size 256 you will likely see some changes in mAP, since the training is doing 4X fewer updates.
> 
> This means that the largest models will probably benefit the most from multi-gpu as they can grow their batch size towards 64, while the smallest model is already capable of batch size 64 with single-gpu.


My highest priority is to align the performance (to boost it), so I have been testing the code with a total batch size of 64 on YOLOv5s.
Theoretically, DDP with batch size 64 and gpu 4 should have similar performance with single gpu with batch size 64 (4 accumulations).
My recent expriment still fails. I will take a try on SyncBN!

Under my setting (DDP with batch size 64 and gpu 4 on v5s), training is 2.0x faster than single-gpu. DP is approximately 1.5x. The DDP acceleration is lower than expected, probably partly due to the small size of model.

**Comment by glenn-jocher** [comment]
@MagicFrogSJTU perfect, yes batch-size 64 in your tests should then produce the same result as the official weights!

Ah, so the acceleration is better than the default code? Definitely next step is to repeat the experiment with SyncBN.

**Comment by MagicFrogSJTU** [comment]
> @MagicFrogSJTU perfect, yes batch-size 64 in your tests should then produce the same result as the official weights!
> 
> Ah, so the acceleration is better than the default code? Definitely next step is to repeat the experiment with SyncBN.

Yeah. It only makes sense if we can reproduce the performance!
I am trying!

**Comment by MagicFrogSJTU** [comment]
> @MagicFrogSJTU perfect, yes batch-size 64 in your tests should then produce the same result as the official weights!
> 
> Ah, so the acceleration is better than the default code? Definitely next step is to repeat the experiment with SyncBN.

I was expecting 3.0-3.5x acceleration. Now it is 2.0x. If the performance issue is resolved I will then try to fix the speed.

**Comment by MagicFrogSJTU** [comment]
It's kind of weird.
After SyncBN integrated, the DDP mode still doesn't work.
Run in DDP with 1 GPU, that's okay.
Run in DDP with 2 GPUs, worse 
Run in DDP with 4 GPUs, more worse.
Kind of weird.

**Comment by NanoCode012** [comment]
Mine is still running, but so far I find that apex.parallel is faster than torch.nn.parallel by quite a bit. Do we need to adjust hyp parameters like learning rate?

**Comment by MagicFrogSJTU** [comment]
> Mine is still running, but so far I find that apex.parallel is faster than torch.nn.parallel by quite a bit. Do we need to adjust hyp parameters like learning rate?

As SyncBN is introduced, the performance of the first 1 epoch if DDP should be the same as that of single-gpu mode. You don't have to train a lot of epochs.
By the way, I have tried removing the ModelEma, and performance result did not change.
apex is faster? But it doesn't matter now as both can't get correct training performance.
This is very weird.

**Comment by MagicFrogSJTU** [comment]
> Mine is still running, but so far I find that apex.parallel is faster than torch.nn.parallel by quite a bit. Do we need to adjust hyp parameters like learning rate?

Learning rate should be the same. If DDP is correctly set, theoretically, We don't have to change the hyperparameter unless it is required to keep DDP right.

**Comment by NanoCode012** [comment]
> > Mine is still running, but so far I find that apex.parallel is faster than torch.nn.parallel by quite a bit. Do we need to adjust hyp parameters like learning rate?
> 
> Learning rate should be the same. If DDP is correctly set, theoretically, We don't have to change the hyperparameter unless it is required to keep DDP right.

I see. I was just suggesting it as an idea because of a thread I read.
https://discuss.pytorch.org/t/training-performance-degrades-with-distributeddataparallel/47152/15

**Comment by MagicFrogSJTU** [comment]
> > > Mine is still running, but so far I find that apex.parallel is faster than torch.nn.parallel by quite a bit. Do we need to adjust hyp parameters like learning rate?
> > 
> > 
> > Learning rate should be the same. If DDP is correctly set, theoretically, We don't have to change the hyperparameter unless it is required to keep DDP right.
> 
> I see. I was just suggesting it as an idea because of a thread I read.
> https://discuss.pytorch.org/t/training-performance-degrades-with-distributeddataparallel/47152/15

You can have a try.
I have done something like this, tried a bigger learning rate. The perf get better in the very first, but gradually became worse than that of single-gpu. But as SyncBN is introduced now, it might change.

**Comment by MagicFrogSJTU** [comment]
It has been fixed in #401
